src/bindings/io/doc/io_doc.c

```diff
@@ -0,0 +1,365 @@
+}
+
+/**
+ * Remove timer from IO-Binding
+ * @see kdbio.h ElektraIoInterface::removeTimer
+ */
+static int ioDocBindingRemoveTimer (ElektraIoTimerOperation * timerOp)
+{
+	assert (elektraIoTimerGetBindingData (timerOp) != NULL);
+	DocOperationData * operationData = (DocOperationData *)elektraIoTimerGetBindingData (timerOp);
+
+	// Stop polling of the file descriptor and free data afterwards
+	// someIoLibStopTimer (operationData->handle);
+	elektraFree (operationData);
+
+	return 0;
+}
+
+/**
+ * Update idle operation in IO-Binding
+ * @see kdbio.h ElektraIoInterface::updateIdle
+ */
+static int ioDocBindingUpdateIdle (ElektraIoIdleOperation * idleOp)
+{
+	assert (elektraIoIdleGetBindingData (idleOp) != NULL);
+	DocOperationData * operationData = (DocOperationData *)elektraIoIdleGetBindingData (idleOp);
+
+	// Based on elektraIoIdleIsEnabled (idleOp) enable or disable the operation
+	// someIoLibUpdateIdle (operationData->handle, elektraIoIdleIsEnabled (idleOp));
+
+	return 0;
+}
+
+/**
+ * Add idle operation to IO-Binding
+ * @see kdbio.h ElektraIoInterface::addIdle
+ */
+static int ioDocBindingAddIdle (ElektraIoInterface * binding, ElektraIoIdleOperation * idleOp)
+{
+	DocOperationData * operationData = newOperationData ();
+	if (operationData == NULL)
+	{
+		return -1;
+	}
+	// You can use private data stored in the IO binding
+	// e.g. MyData data = (MyData *)elektraIoBindingGetData (binding);
+
+	elektraIoIdleSetBindingData (idleOp, operationData);
+	// You can store additional data for each operation in your operationData structure
+	operationData->operation.idle = idleOp;
+
+	// Here you need to add the operation to the I/O management library
+	// ioDocBindingIdleCallback holds an example callback to pass to your I/O management library
+
+	// assume SomeIoLibHandle * someIoLibAddIdle(int enabled, void * privateData, callback)
+	// operationData->handle = someIoLibAddIdle(elektraIoIdleIsEnabled (idleOp), &idleOp, ioDocBindingIdleCallback)
+
+	return 0;
+}
+
+/**
+ * Remove idle operation from IO-Binding
+ * @see kdbio.h ElektraIoInterface::removeIdle
+ */
+static int ioDocBindingRemoveIdle (ElektraIoIdleOperation * idleOp)
+{
+	assert (elektraIoIdleGetBindingData (idleOp) != NULL);
+	DocOperationData * operationData = (DocOperationData *)elektraIoIdleGetBindingData (idleOp);
+
+	// Stop polling of the file descriptor and free data afterwards
+	// someIoLibStopIdle (operationData->handle);
+	elektraFree (operationData);
+
+	return 0;
+}
+
+/**
+ * Cleanup
+ * @param  binding IO-Binding
+ * @see kdbio.h ElektraIoInterface::cleanup
+ */
+static int ioDocBindingCleanup (ElektraIoInterface * binding)
+{
+	assert (binding != NULL);
+	elektraFree (elektraIoBindingGetData (binding));
+	elektraFree (binding);
+	return 0;
+}
+
+/**
+ * Create and initialize a new doc IO binding
+ * @param  foo Some data from I/O management library (e.g. a handle)
+ * @return     Populated I/O interface
+ */
+ElektraIoInterface * elektraIoDocNew (char * foo)
```

Comment:
Take a look at how ELEKTRA_PLUGIN_EXPORT works (src/include/kdbplugin.h). Even if we do not introduce the static/full variants for I/O bindings, code using ELEKTRA_IO_BINDING_EXPORT would be more compact and declarative.

Reply:
I see. Varargs would make the code more declarative and compact but they sacrifice type safety. Using a struct (a struct just for this purpose) could be the middle ground, while the current solution is very compact but not declarative.

For plugins varargs provide the additional advantage of being able to omit optional functions, which is currently not needed. 
But if a new (and optional) function was necessary for I/O bindings we can maintain ABI compatibility. Maintaining ABI compatibility for structs seems [quite painful](https://stackoverflow.com/questions/25947997/how-does-sizeofstruct-help-provide-abi-compatibility) compared to the varargs solution. With functions ABI compatibility holds only as long as the prototype stays the same.

In summary I would trade type safety for possible ABI compatibility. What do you think?

Reply:
If you do not need optional parameters having everything as parameters sounds okay. Other differences of ELEKTRA_PLUGIN_EXPORT are:

1. allocation hidden within the function
2. the method name (elektraIoDocNew) is hidden by the macro. Doing so allows us to have static linkage and variants.

Reply:
Okay, thank you.

Allocation is already done in `elektraIoNewBinding`. Other allocations are specific to bindings, but could be moved to an additional "open" function (like done with plugins).
Hiding the method name also comes with additional work when loading the binding (e.g. dl.c / static.c). 

For now there I see little reason for static linking or variantes. Since the interface is experimental we can still change this later.

Reply:
Ahh, I see. I was confused about that elektraIoDocNew also implements the "Open" of the binding.

If there is no reason to separate these concepts it can stay as it is.

Reply:
Ok, great! :)